### PR TITLE
Implement `TargetRegistry`.

### PR DESCRIPTION
**Pull request by jmalloc**
This PR implements a registry of targets. A target, in gRPC parlance is a bit like a DSN for a DBMS - it describes some endpoint that can be connected to.  Typically it is a hostname and port, but not necessarily.

The registry is an observable collection of the targets that have been found by some as-yet- undescribed discovery process.

### CODE REVIEW

File: target.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	r.m.RLock()
+	defer r.m.RUnlock()
+
+	_, ok := r.targets[t]
+	return ok
+}
+
+// Select returns a random target in the given group.
+//
+// If there are targets in the given group, ok is false.
```
Comment (koden-km) [inline target.go]: I think you mean if there are **no** targets then ok is false?
```suggestion
// If there are no targets in the given group, ok is false.
```

File: target.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+// Has returns true if t is in the registry.
+func (r *TargetRegistry) Has(t *Target) bool {
+	r.m.RLock()
+	defer r.m.RUnlock()
+
+	_, ok := r.targets[t]
+	return ok
+}
+
+// Select returns a random target in the given group.
```
Comment (koden-km) [inline target.go]: Just to confirm, is this funciton returning a random target by using Go's automatic random map iteration order?

File: target_test.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+				defer GinkgoRecover()
+				Fail("observer unexpectedly notified of the same target")
+			}
+
+			reg.Add(target1)
+		})
+	})
+
+	Describe("func Remove()", func() {
+		It("notifies the observers of the removing the target", func() {
```
Comment (koden-km) [inline target_test.go]: There is something wrong with this sentence.

File: target.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+//
+// If there are targets in the given group, ok is false.
+func (r *TargetRegistry) Select(g string) (t *Target, ok bool) {
+	r.m.RLock()
+	defer r.m.RUnlock()
+
+	for t := range r.groups[g] {
+		return t, true
```
Comment (danilvpetrov) [inline target.go]: I don't seem to understand this line of code. It is supposed to return `map[*Target]struct{}` but the return type is `*Target`.

File: target.go (hunk 2 above)
Comment (jmalloc) [reply target.go]: Yep

File: target_test.go (hunk 3 above)
Comment (jmalloc) [reply target_test.go]: Yes :)

File: target.go (hunk 4 above)
Comment (jmalloc) [reply target.go]: `r.groups` is a map-of-maps, so `r.groups[g]` is a map itself, and I iterate and return immediately just to exploit the guaranteed randomisation of `range`.

File: target.go (hunk 2 above)
Comment (koden-km) [reply target.go]: I wonder if that should be commented just to make it clear? Maybe you Go guys are used to that pattern though.

File: target.go (hunk 4 above)
Comment (danilvpetrov) [reply target.go]: ok, that's more clear. But is range `random` or `arbitrary`? IMHO, the same map will return the same item repeatedly without any randomization, it's just in what order the items are stored is not guaranteed.

Comment (jmalloc) [reply target.go]: You might be right. I do recall reading that it was *intentionally* randomised, but searching for it now I can't confirm that.

Comment (jmalloc) [reply target.go]: Ok, I've changed it to use explicit randomisation of a slice.

**Pull request merged**